File: java/tournament/src/main/java/Tournament.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import java.util.stream.Collectors;

class Tournament {

    private Map<String, Team> teams;

    public Tournament() {
        this.teams = new HashMap<>();
    }

    String printTable() {
        String header = "Team                           | MP |  W |  D |  L |  P";

        String table = this.teams.values().stream()
                .sorted().map(this::printTeam).collect(Collectors.joining("\n"));

        return header + "\n" + (table.isBlank() ? "" : table + "\n");
    }

    void applyResults(String resultString) {
        Arrays.stream(resultString.split("\n")).forEach(this::applyResult);
    }

    private void applyResult(String matchResult) {
        String[] match = matchResult.split(";");

        String teamA = match[0];
        String teamB = match[1];
        String result = match[2];

        Team a = this.teams.getOrDefault(teamA, new Team(teamA));
        Team b = this.teams.getOrDefault(teamB, new Team(teamB));
        a.addMatch();
        b.addMatch();

        if (result.equals("win")) {
            a.addWin();
            b.addLoss();
        } else if (result.equals("loss")) {
            a.addLoss();
            b.addWin();
        } else {
            a.addDraw();
            b.addDraw();
        }

        teams.put(teamA, a);
        teams.put(teamB, b);
    }

    private String printTeam(Team team) {
        return String.format("%s|%s |%s |%s |%s |%s",
                String.format("%-31s", team.getName()), 
                String.format("%3d", team.getMatches()), 
                String.format("%3d", team.getWins()), 
                String.format("%3d", team.getDraws()), 
                String.format("%3d", team.getLosses()), 
                String.format("%3d", team.getPoints()));
    }
}

File: javascript/tournament/tournament.js
export const tournamentTally = (input) => {

    let header = 'Team                           | MP |  W |  D |  L |  P';
    if (!input) {
        return header;
    }

    return header + '\n' + 
        Object.entries(input.split('\n').reduce(parseGame, {}))
                .toSorted(sortingFunction)
                .map(toString) .join('\n');

};

const parseGame = (tournament, game) => {
    let [teamA, teamB, result] = game.split(';')

    let stateA = tournament[teamA] ? tournament[teamA] : { mp: 0, w: 0, d: 0, l: 0 , p: 0};
    let stateB = tournament[teamB] ? tournament[teamB] : { mp: 0, w: 0, d: 0, l: 0 , p: 0};
    stateA.mp++;
    stateB.mp++;

    switch(result) {
        case 'win':
            stateA.w++;
            stateA.p += 3;
            stateB.l++;
            break;
        case 'draw':
            stateA.d++;
            stateA.p++;
            stateB.d++;
            stateB.p++;
            break;
        case 'loss':
            stateA.l++;
            stateB.w++;
            stateB.p += 3;
            break;
    }

    return {...tournament, [teamA]: stateA, [teamB]: stateB};

};

const sortingFunction = ([nameA, stateA], [nameB, stateB]) => {
    if (stateA.p === stateB.p) {
        return nameA < nameB ? -1 : nameA > nameB ? 1 : 0;
    }

    return stateB.p - stateA.p;
};

const toString = ([teamName, {mp, w, d, l, p}]) => {
    let teamHeaderLength = 'Team                           '.length;
    let name = teamName.padEnd(teamHeaderLength, ' ');
    let points = p.toString().padStart(3, ' ');

    return `${name}|  ${mp} |  ${w} |  ${d} |  ${l} |${points}`;
};
